packages/commonwealth/server/util/roles.ts

```diff
@@ -0,0 +1,262 @@
+import { Model, Transaction, Op, FindOptions } from 'sequelize';
+import { DB } from 'server/models';
+import { sequelize } from '../database';
+import { Permission } from '../models/role';
+import { RoleAssignmentAttributes } from '../models/role_assignment';
+
+export class RoleInstanceWithPermission {
+  _roleAssignmentAttributes: RoleAssignmentAttributes;
+  chain_id: string;
+  permission: Permission;
+  allow: bigint;
+  deny: bigint;
+
+  constructor(
+    _roleAssignmentInstance: RoleAssignmentAttributes,
+    chain_id: string,
+    permission: Permission,
+    allow: bigint,
+    deny: bigint
+  ) {
+    this._roleAssignmentAttributes = _roleAssignmentInstance;
+    this.chain_id = chain_id;
+    this.permission = permission;
+    this.allow = allow;
+    this.deny = deny;
+  }
+
+  public toJSON(): RoleAssignmentAttributes & {
+    chain_id: string;
+    permission: Permission;
+    allow: bigint;
+    deny: bigint;
+  } {
+    return {
+      ...this._roleAssignmentAttributes,
+      chain_id: this.chain_id,
+      permission: this.permission,
+      allow: this.allow,
+      deny: this.deny,
+    };
+  }
+}
+
+export async function findAllRoles(
+  models: DB,
+  findOptions: FindOptions,
+  chain_id?: string,
+  permissions?: Permission[]
+): Promise<RoleInstanceWithPermission[]> {
+  let roleFindOptions: any;
+  const whereCondition = {};
+  if (chain_id) {
+    whereCondition['chain_id'] = chain_id;
+  }
+  if (permissions) {
+    roleFindOptions = {
+      where: {
+        [Op.and]: [
+          { chain_id },
+          {
+            name: {
+              [Op.or]: permissions,
+            },
+          },
+        ],
+      },
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+    };
+  } else {
+    roleFindOptions = {
+      where: whereCondition,
+      include: [
+        {
+          model: models.RoleAssignment,
+          ...findOptions,
+        },
+      ],
+    };
+  }
+  const communityRoles = await models.CommunityRole.findAll(roleFindOptions);
+  const roles: RoleInstanceWithPermission[] = [];
+  if (communityRoles) {
+    for (const communityRole of communityRoles) {
+      const roleAssignments = communityRole.RoleAssignments;
+      if (roleAssignments.length > 0) {
+        for (const roleAssignment of roleAssignments) {
+          const role = new RoleInstanceWithPermission(
+            roleAssignment.toJSON(),
+            communityRole.chain_id,
+            communityRole.name,
+            communityRole.allow,
+            communityRole.deny
+          );
+          roles.push(role);
+        }
+      }
+    }
+  } else {
+    return null;
+  }
+  return roles;
+}
+
+export async function findOneRole(
+  models: DB,
+  findOptions: FindOptions,
+  chain_id: string,
+  permissions?: Permission[]
+): Promise<RoleInstanceWithPermission> {
+  let roleFindOptions: any;
+  if (permissions) {
```

Comment:
Why so much duplicate code here, from above call? Can't we reuse `findAllRoles` and synthesize a simpler result?